CASSGO-5 Fix DisableInitialHostLookup flag ignored when querying system.peers

RostislavPorohnya opened the pull request:
This PR addresses issue #1665, which reports a bug where `system.peers` table is queried even if `DisableInitHostLookup` flag is set to `true`. According to the driver documentation, `system.peers` should not be queried when the flag is set to `true`

In this PR `DisableInitialHostLookup` flag renamed to `DisableHostLookup`, code logic is fixed to query `system.peers` only when the flag is set to `false`.

RostislavPorohnya committed:
Modified DisableInitialHostLookup flag name and logic

joao-r-reis commented on conn.go:
```diff
@@ -1689,36 +1689,37 @@ func (c *Conn) querySystemLocal(ctx context.Context) *Iter {
+
+			if err = iter.Close(); err != nil {
+				goto cont
+			}
```
I think it's fine if we query system.peers for the purpose of retrieving the schema version of each host as long as we don't update gocql's ring/hosts. If a user wants to use a static set of contact points and avoid using hosts from system tables there's no reason why they would have to give up the functionality of awaiting for schema agreement.

tengu-alt replied:
Agree, fixed.

joao-r-reis commented on host_source.go:
```diff
@@ -635,6 +635,11 @@ func (r *ringDescriber) getClusterPeerInfo(localHost *HostInfo) ([]*HostInfo, er
+	// Check if host lookup is disabled
+	if r.session.cfg.DisableHostLookup {
+		return []*HostInfo{}, nil
+	}
```
Instead of adding this, I think we should make the `ringRefresher` do a NOOP when this flag is set. `getClusterPeerInfo` should only be about retriving the peers info from the server so it should be up to the caller whether this should be done or not.

If we make the `func refreshRing(r *ringDescriber) error` function do nothing when the flag is set then it should solve all of the issues related to the driver not respecting the flag.

tengu-alt replied:
Agree, refactored

joao-r-reis commented on cluster.go:
```diff
@@ -189,12 +189,12 @@ type ClusterConfig struct {
 	// mean that data_centre, rack and token information will not be available and as
 	// such host filtering and token aware query routing will not be available.
-	DisableInitialHostLookup bool
+	DisableHostLookup bool
```
I don't think this is worth breaking applications. I agree that removing the `Initial` word from the name does make sense but it's really not that big of a deal.

joao-r-reis replied:
At the very least we would have to keep both flags (+ deprecate the old one) and map it to a single internal flag so that we can do a "safe" rename of the flag without breaking user applications.

tengu-alt replied:
Agree, I've returned the old name.

joao-r-reis reviewed (commented):
Added some comments.

In summary I think this PR should be limited to making the `refreshRing` function do nothing when the flag is set. That would ensure that the driver will keep using the contact points as the hosts forever and ignore any kind of topology changes or any host information from system tables. 

It's fine if the driver queries the system tables to check if there is schema agreement, I don't think it goes against the intention of the flag.

OleksiienkoMykyta reviewed (commented):
It looks like your PR helps with (https://github.com/apache/cassandra-gocql-driver/issues/1789) and (https://github.com/apache/cassandra-gocql-driver/issues/1665), but what is the status of it? Should it be updated according to comments, or closed in favor of some other PR?

joao-r-reis commented:
If there's no update to this PR soon I'll descope it from the 2.0 release as this can be released in a minor released without an issue

RostislavPorohnya commented:
@joao-r-reis I'm going to update it by the end of next week, thanks for the comments!

RostislavPorohnya commented:
Hi. I’ve handed this PR over to @tengu-alt, who will finalize the remaining work. I’m switching to another task, so please feel free to tag him if you have any questions or need updates.

tengu-alt commented:
Made some updates: refactored tests, added new logic(`refreshRing()` from now does NOOP when `DisableInitialHostLookup` is true), and refactored code according to feedback. Also fixed the conflicts .

tengu-alt commented:
@joao-r-reis Could you please take a look?

joao-r-reis commented on CHANGELOG.md:
```diff
@@ -50,6 +50,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
+- Fix DisableInitialHostLookup flag ignored when querying system.peers (CASSGO-5)
+
+- Fix DisableInitialHostLookup flag ignored when querying system.peers (PR update) (CASSGO-5)
```
duplicate entry?

tengu-alt replied:
Agree, deleted

joao-r-reis commented on conn.go:
```diff
@@ -1898,16 +1898,14 @@ func (c *Conn) querySystemLocal(ctx context.Context) *Iter {
 	for time.Now().Before(endDeadline) {
 		iter := c.querySystemPeers(ctx, c.host.version)
 
-		versions = make(map[string]struct{})
```
why this change? resetting the map on each attempt made more sense to me

tengu-alt replied:
Agree, fixed

joao-r-reis commented on disable_host_lookup_test.go:
```diff
@@ -0,0 +1,39 @@
+	"testing"
+)
+
+func TestDisableHostLookupRingRefresh(t *testing.T) {
```
can we not add this to an existing test file? Creating a new file just for this test seems excessive. Also this is running in "unit tests" step because it's missing the appropriate build flag at the top of the file. Also missing license header.

tengu-alt replied:
Done, I've moved the test into the `cassandra_test.go`

joao-r-reis commented:
Also, I'm considering pushing this to a 2.x minor release followup and descoping it from 2.0 because it doesn't really need a major release and I also want do some extensive testing with the `DisableInitialHostLookup` flag as I believe there's some other issues that still need to be fixed (e.g. system.local still being queried I think)

tengu-alt committed:
Fix DisableInitialHostLookup flag ignored when querying system.peers(PR update)

This commit provides update for the 'Modified DisableInitialHostLookup flag name and logic'
refreshRing() does NOOP when 'DisableInitialHostLookup' is true. system.peers
can be queried when flag is true for the schema version retrieving.

patch by Oleksandr Luzhniy; for CASSGO-5
